File: packages/rayact-linking/ios/RayactLinkingRegistration.swift
import Foundation
import UIKit

final class RayactLinkingRegistration: RayactPlatformModuleRegistration {
    func register(with registry: RayactPlatformRegistry) {
        registry.registerModule("linking") { method, payload, completion in
            let value = (try? JSONSerialization.jsonObject(with: payload))
                as? [String: Any] ?? [:]
            guard let rawURL = value["url"] as? String, let url = URL(string: rawURL) else {
                completion(.failure(NSError(
                    domain: "RayactLinking",
                    code: 1,
                    userInfo: [NSLocalizedDescriptionKey: "A valid URL is required"]
                )))
                return
            }
            let perform = {
                switch method {
                case "canOpenURL":
                    completion(.success(Data((UIApplication.shared.canOpenURL(url) ? "true" : "false").utf8)))
                case "openURL":
                    guard UIApplication.shared.canOpenURL(url) else {
                        completion(.failure(NSError(
                            domain: "RayactLinking",
                            code: 2,
                            userInfo: [NSLocalizedDescriptionKey: "No application can open URL: \(rawURL)"]
                        )))
                        return
                    }
                    UIApplication.shared.open(url, options: [:]) { opened in
                        if opened {
                            completion(.success(Data("true".utf8)))
                        } else {
                            completion(.failure(NSError(
                                domain: "RayactLinking",
                                code: 3,
                                userInfo: [NSLocalizedDescriptionKey: "Unable to open URL: \(rawURL)"]
                            )))
                        }
                    }
                default:
                    completion(.failure(NSError(
                        domain: "RayactLinking",
                        code: 4,
                        userInfo: [NSLocalizedDescriptionKey: "Unknown linking method '\(method)'"]
                    )))
                }
            }
            if Thread.isMainThread {
                perform()
            } else {
                DispatchQueue.main.async(execute: perform)
            }
        }
    }
}

File: packages/rayact-linking/android/src/main/java/dev/rayact/linking/RayactLinkingRegistration.kt
package dev.rayact.linking

import android.content.Intent
import android.net.Uri
import com.rayact.engine.RayactPlatformModuleRegistration
import com.rayact.engine.RayactPlatformRegistry
import org.json.JSONObject

class RayactLinkingRegistration : RayactPlatformModuleRegistration {
    override fun register(
        context: android.content.Context,
        registry: RayactPlatformRegistry,
    ) {
        registry.registerModule("linking") { method, payload, completion ->
            runCatching {
                val data = runCatching { JSONObject(payload.toString(Charsets.UTF_8)) }
                    .getOrElse { JSONObject() }
                val url = data.optString("url")
                require(url.isNotBlank()) { "A URL is required" }
                val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
                when (method) {
                    "canOpenURL" -> {
                        val available = intent.resolveActivity(context.packageManager) != null
                        available.toString().toByteArray(Charsets.UTF_8)
                    }
                    "openURL" -> {
                        val activity = RayactPlatformRegistry.currentActivity()
                            ?: error("No foreground activity is available")
                        require(intent.resolveActivity(context.packageManager) != null) {
                            "No application can open URL: $url"
                        }
                        activity.runOnUiThread { activity.startActivity(intent) }
                        "true".toByteArray(Charsets.UTF_8)
                    }
                    else -> error("Unknown linking method '$method'")
                }
            }.fold(
                onSuccess = { completion(Result.success(it)) },
                onFailure = { completion(Result.failure(it)) },
            )
        }
    }
}
